### PR TITLE
add dag_factory

### PR DESCRIPTION
**Pull request by CaptainOfHacks**

### CODE REVIEW

**Review by Dragos0000 — APPROVED** [review]

**Review by DanCh11 — APPROVED** [review]

File: sem_covid/adapters/dag_factory.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import abc
+from abc import abstractmethod
+
+from airflow import DAG
+from airflow.operators.python import PythonOperator
+
+
+class DagPipeline(abc.ABC):
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:8]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    @abstractmethod
+    def get_steps(self) -> list:
+        pass
+
+
+class DagStep:
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:15]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self, dag_pipeline: DagPipeline, dag_pipeline_step):
+        self.dag_pipeline = dag_pipeline
+        self.dag_pipeline_step = dag_pipeline_step
+
+    @abstractmethod
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+class ObjectStateManager(abc.ABC):
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:25]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    @abstractmethod
+    def save_object_state(self, obj: object):
+        pass
+
+    @abstractmethod
+    def load_object_state(self) -> object:
+        pass
+
+
+class DagPipelineManager:
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:36]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    def create_step(self, dag_pipeline_step, stateful: bool = False) -> DagStep:
+        if stateful:
+            assert self.object_state_manager is not None
+            return StatefulDagStep(self.dag_pipeline, dag_pipeline_step, self.object_state_manager)
+        else:
+            return StatelessDagStep(self.dag_pipeline, dag_pipeline_step)
+
+
+class StatefulDagStep(DagStep):
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:49]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        getattr(self.dag_pipeline, self.dag_pipeline_step.__name__)(*args, **kwargs)
+        self.object_state_manager.save_object_state(self.dag_pipeline)
+
+
+class StatelessDagStep(DagStep):
+    def __call__(self, *args, **kwargs):
+        getattr(self.dag_pipeline, self.dag_pipeline_step.__name__)(*args, **kwargs)
+
+
+class DagFactory:
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:65]: Add documentation how this shall be used

File: sem_covid/adapters/dag_factory.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        super().__init__(dag_pipeline, dag_pipeline_step)
+        self.object_state_manager = object_state_manager
+
+    def __call__(self, *args, **kwargs):
+        self.dag_pipeline = self.object_state_manager.load_object_state()
+        getattr(self.dag_pipeline, self.dag_pipeline_step.__name__)(*args, **kwargs)
+        self.object_state_manager.save_object_state(self.dag_pipeline)
+
+
+class StatelessDagStep(DagStep):
```
Comment (costezki) [inline sem_covid/adapters/dag_factory.py:60]: Add documentation how this shall be used

**Review by costezki — COMMENTED** [review]
Looks nice! Still a bit more complex than I was imagining but I guess that is for a reason. 
Well done guys!

**Pull request merged**